extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -1676,4 +1700,88 @@ public void run()
+                  new Function<Map<Integer, Long>, Void>()
+                  {
+                    @Override
+                    public Void apply(Map<Integer, Long> taskResponse)
+                    {
+                      if (taskResponse != null) {
+                        for (final Map.Entry<Integer, Long> partitionOffsets : taskResponse.entrySet()) {
+                          offsetsResponse.compute(partitionOffsets.getKey(), new BiFunction<Integer, Long, Long>()
+                          {
+                            @Override
+                            public Long apply(Integer key, Long existingOffsetInMap)
+                            {
+                              // If existing value is null use the offset returned by task
+                              // otherwise use the max (makes sure max offset is taken from replicas)
+                              return existingOffsetInMap == null
+                                     ? partitionOffsets.getValue()
+                                     : Math.max(partitionOffsets.getValue(), existingOffsetInMap);
+                            }
+                          });
+                        }
+                      }
+                      return null;
+                    }
+                  })
+              );
+            }
+          }
+          // not using futureTimeoutInSeconds as its min value is 120 seconds and this metric is emitted every minute
+          Futures.successfulAsList(futures).get(30, TimeUnit.SECONDS);
+
+          // for each partition, seek to end to get the highest offset
+          // check the offsetsResponse map for the latest consumed offset
+          // if partition info not present in offsetsResponse then use lastCurrentOffsets map
+          // if not present there as well, fail the compute
+
+          long lag = 0;
+          for (PartitionInfo partitionInfo : partitionInfoList) {
+            final TopicPartition topicPartition = new TopicPartition(ioConfig.getTopic(), partitionInfo.partition());
+            lagComputingConsumer.seekToEnd();
+            if (offsetsResponse.get(topicPartition.partition()) != null) {
+              lag += (lagComputingConsumer.position(topicPartition) - offsetsResponse.get(topicPartition.partition()));
+              lastCurrentOffsets.put(topicPartition.partition(), offsetsResponse.get(topicPartition.partition()));
+            } else if (lastCurrentOffsets.get(topicPartition.partition()) != null) {
+              lag += (lagComputingConsumer.position(topicPartition)
+                      - lastCurrentOffsets.get(topicPartition.partition()));
+            } else {
+              throw new ISE("Could not find latest consumed offset for partition [%d]", topicPartition.partition());
+            }
+          }
+          emitter.emit(
+              ServiceMetricEvent.builder().setDimension("dataSource", dataSource).build("ingest/kafka/lag", lag)
+          );
```

Comment:
In case there are issues with specific kafka partitions, knowledge about lag of individual partitions might be useful. Not a blocker for this PR.